addons/account_edi/models/account_move.py

```diff
@@ -135,6 +135,66 @@ def _compute_edi_show_abandon_cancel_button(self):
     # Export Electronic Document
     ####################################################
 
+    def _prepare_edi_vals_to_export(self):
+        ''' The purpose of this helper is to prepare values in order to export an invoice through the EDI system.
+        This includes the computation of the tax details for each invoice line that could be very difficult to
+        handle regarding the computation of the base amount.
+        :return: A python dict containing default pre-processed values.
+        '''
+        self.ensure_one()
+
+        res = {
+            'record': self,
+            'invoice_line_vals_list': [],
+        }
+
+        # Invoice lines details.
+        tax_detail_per_tax = {}
+        added_base_amount_keys = set()
+        for index, line in enumerate(self.invoice_line_ids.filtered(lambda line: not line.display_type), start=1):
+            line_vals = {
+                'index': index,
+                **line._prepare_edi_vals_to_export(),
+            }
+            res['invoice_line_vals_list'].append(line_vals)
+
+            # Tax details.
+            for tax_vals in line_vals['tax_detail_vals_list']:
+                tax_detail_per_tax.setdefault(tax_vals['tax'], {
+                    'tax': tax_vals['tax'],
+                    'src_tax': tax_vals['src_tax'],
+                    'tax_base_amount_currency': 0.0,
+                    'tax_amount_currency': 0.0,
+                    'tag_ids': set(),
+                })
+                vals = tax_detail_per_tax[tax_vals['tax']]
+
+                # Avoid adding multiple times the same base (e.g. with multiple taxes on the same line).
+                base_amount_key = (line.id, tax_vals['tax']['id'])
+                if base_amount_key not in added_base_amount_keys:
+                    vals['tax_base_amount_currency'] += tax_vals['tax_base_amount_currency']
+                    added_base_amount_keys.add(base_amount_key)
+
+                vals['tax_amount_currency'] += tax_vals['tax_amount_currency']
+                for tag in tax_vals['tags']:
+                    vals['tag_ids'].add(tag.id)
+
+        # Format the aggregated tax details as a list.
+        res['tax_detail_vals_list'] = []
+        for tax_detail_vals in tax_detail_per_tax.values():
+            res['tax_detail_vals_list'].append({
+                **tax_detail_vals,
+                'tags': self.env['account.account.tag'].browse(tax_detail_vals['tag_ids']),
+            })
+
+        # Totals.
+        res.update({
+            'total_price_subtotal_wo_discount': sum(x['price_subtotal_wo_discount'] for x in res['invoice_line_vals_list']),
```

Comment:
maybe total_price_subtotal_before_discount is clearer.

Reply:
Idk... you are suggesting `before` in this comment but `after` in the next one. I'm suggesting `wo` for `without` but none of these suggestions are optimal and very clear... to discuss

Reply:
@smetl Because it is different:
price_unit is by default before discount and the price_subtotal field of the account.move is after discount.  And you are doing price_unit_after_discount and price_subtotal_before_discount (before applying the discount)